File: #131_Palindrome Partition/Solution.java
import java.util.List;
import java.util.Deque;
import java.util.LinkedList;

/* backtracing takes O(2^N) time
 * could be slightly imporved by using dp to keep track of palindrome 
 */

public class Solution {
    public static List<List<String>> partition(String s) {
        if (s == null || s.length() == 0) return null;
        List<List<String>> result = new LinkedList<List<String>>();
        LinkedList<String> list = new LinkedList<String>();
        partition(s, result, list);
        return result;
    }

    private static void partition(String s, List<List<String>> result, LinkedList<String> list) {
    	if (s.length() == 0 || s == null) return;
    	for (int i = 1; i <= s.length(); i++) {
    		if (isPalindrome(s.substring(0, i))) {
    			list.add(s.substring(0, i));
    			if (i == s.length()) {
    				result.add(new LinkedList<String>(list));
    			}
    			else partition(s.substring(i), result, list);
    			list.removeLast();
    		}
    	}
    }

    private static boolean isPalindrome(String s) {
    	if (s == null || s.length() == 0) return false;
    	if (s.length() == 1) return true;
    	int len = s.length();
    	int start = 0; int end = len - 1;
    	while (true) {
    		if (s.charAt(start) != s.charAt(end)) return false;
    		start++; end--;
    		if (start >= end) break;
    	}
    	return true;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	List<List<String>> result = new LinkedList<List<String>>();
    	result = sol.partition("aab");
    	// System.out.println(result.size());
    	for (List<String> list : result) {
    		for (String str : list) {
    			System.out.print(str + " ");
    		}
    		System.out.println();
    	}
    }
}

File: #131_Palindrome Partition/Solution.py
'''
Use dp to see if s[i, j] is palindrome in O(n ^ 2)
Then backtracing
'''

class Solution:

	def __init__(self):
		self.dp = None
		self.ans = []

	# @param {string} s
	# @return {string[][]}
	def partition(self, s):
		if not s:
			return []
		self.dp = [[False for j in range(len(s))] for i in range(len(s))]
		self.ans = []
		for x in range(len(s)):
			self.dp[x][x] = True
		for j in range(len(s)):
			for i in range(len(s)):
				if i >= j:
					break
				elif j - i == 1 and s[i] == s[j]:
					self.dp[i][j] = True
				elif s[i] == s[j] and self.dp[i + 1][j - 1]:
					self.dp[i][j] = True
		self.backtrace(s, 0, [])
		return self.ans

	def backtrace(self, s, index, tmp):
		if index == len(s):
			self.ans.append(list(tmp))
			return
		for x in range(index, len(s)):
			if self.dp[index][x]:
				tmp.append(s[index : x + 1])
				self.backtrace(s, x + 1, tmp)
				tmp.pop()

sol = Solution()
print sol.partition("aab")